### PR TITLE
SONARJAVA-1442 FP in S1610 for abstract classes when a method is overriden

### PR DESCRIPTION
**Pull request by mpaladin**

### CODE REVIEW

File: java-checks/src/main/java/org/sonar/java/checks/AbstractClassNoFieldShouldBeInterfaceCheck.java [hunk 1]
```diff
@@ -74,10 +74,14 @@ private static boolean classIsAbstract(ClassTree tree) {
...
-      if (member.is(Tree.Kind.VARIABLE) || (member.is(Tree.Kind.METHOD) && ModifiersUtils.hasModifier(((MethodTree) member).modifiers(), Modifier.PROTECTED))) {
+      if (member.is(Tree.Kind.VARIABLE) || (member.is(Tree.Kind.METHOD) && isProtectedOrOverriding((MethodTreeImpl) member))) {
         return false;
       }
     }
     return true;
   }
+
+  private static boolean isProtectedOrOverriding(MethodTreeImpl member) {
+    return ModifiersUtils.hasModifier(member.modifiers(), Modifier.PROTECTED) || member.isOverriding();
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AbstractClassNoFieldShouldBeInterfaceCheck.java]: member.isOverriding can return null... therefore you have a NPE waiting for you right here.
The problem is a bit deeper than that in fact and was introduced with https://github.com/SonarSource/sonar-java/commit/7c493f55e153fc1a1f5154c8dcf8f5ff5cae9167

File: java-checks/src/main/java/org/sonar/java/checks/AbstractClassNoFieldShouldBeInterfaceCheck.java [hunk 2]
```diff
@@ -74,10 +74,14 @@ private static boolean classIsAbstract(ClassTree tree) {
...
-      if (member.is(Tree.Kind.VARIABLE) || (member.is(Tree.Kind.METHOD) && ModifiersUtils.hasModifier(((MethodTree) member).modifiers(), Modifier.PROTECTED))) {
+      if (member.is(Tree.Kind.VARIABLE) || (member.is(Tree.Kind.METHOD) && isProtectedOrOverriding((MethodTreeImpl) member))) {
         return false;
       }
     }
     return true;
   }
+
+  private static boolean isProtectedOrOverriding(MethodTreeImpl member) {
+    return ModifiersUtils.hasModifier(member.modifiers(), Modifier.PROTECTED) || !Boolean.FALSE.equals(member.isOverriding());
```
Comment (benzonico) [inline java-checks/src/main/java/org/sonar/java/checks/AbstractClassNoFieldShouldBeInterfaceCheck.java]: I guess that this also deserves a fix : https://github.com/SonarSource/sonar-java/blob/master/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L269-269

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isOverriding" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/f9474738add3e0aae5d035de7f513c7c0526074e/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L259) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isOverriding" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/1ee4cf502443a2e368e9baa1095ebef7b1b1f136/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L261) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isOverriding" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/1ee4cf502443a2e368e9baa1095ebef7b1b1f136/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L261) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 3]
```diff
@@ -505,29 +504,48 @@ public MethodJavaSymbol overriddenSymbol() {
...
       List<JavaSymbol> symbols = classType.getSymbol().members().lookup(name);
       for (JavaSymbol overrideSymbol : symbols) {
         if (overrideSymbol.isKind(JavaSymbol.MTH)) {
           MethodJavaSymbol methodJavaSymbol = (MethodJavaSymbol) overrideSymbol;
-          if (canOverride(methodJavaSymbol) && Boolean.TRUE.equals(isOverriding(methodJavaSymbol, classType))) {
-            return methodJavaSymbol;
+          if (canOverride(methodJavaSymbol)) {
+            Boolean overriding = isOverriding(methodJavaSymbol, classType);
+            if (overriding == null) {
+              if (!unknownFound) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: MINOR Refactor this code to not nest more than 4 if/for/while/switch/try statements. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS134)

File: java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java [hunk 4]
```diff
@@ -505,29 +504,48 @@ public MethodJavaSymbol overriddenSymbol() {
...
         }
       }
+      if (unknownFound) {
+        return Symbols.unknownMethodSymbol;
+      }
       return null;
     }
 
     @Nullable
     private MethodJavaSymbol overriddenSymbolFrom(JavaType.ClassJavaType classType) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/JavaSymbol.java]: MAJOR The Cyclomatic Complexity of this method "overriddenSymbolFrom" is 12 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isOverriding" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/8f37d87dbc4a1edbc6ef8d430bf1b4ab5cb06a7a/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L261) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [The Cyclomatic Complexity of this method "isOverriding" is 11 which is greater than 10 authorized.](https://github.com/SonarSource/sonar-java/blob/8f37d87dbc4a1edbc6ef8d430bf1b4ab5cb06a7a/java-squid/src/main/java/org/sonar/java/model/declaration/MethodTreeImpl.java#L261) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Pull request merged**